### PR TITLE
8301580: Error recovery does not clear returnResult

### PR DESCRIPTION
**Pull request by lahodaj**
Considering code like:
```
class C {
    void m
    {
        return;
    }
}
```

`void m` is wrapped in an erroneous tree (which is good), and if/when `Attr` is processing it in `visitErroneous`, it will create a new `Env` to attribute the erroneous part.

But, `Attr` will set a `returnResult` into the `Env`'s info - and that info is shared with the outter `Env`, so there will be a `returnResult` set after `visitErroneous`. `{ return ; }` is interpreted as an initializer, and there should be an error for the `return` there, but this error is missing, due to the set `returnResult`. This will then fail later in `Flow` with an exception:

```
$ javac -XDshould-stop.at=FLOW -XDdev /tmp/C.java
/tmp/C.java:3: error: '(' expected
    {
    ^
1 error
An exception has occurred in the compiler (19.0.1-internal). Please file a bug against the Java compiler via the Java bug reporting page (http://bugreport.java.com/) after checking the Bug Database (http://bugs.java.com/) for duplicates. Include your program, the following diagnostic, and the parameters passed to the Java compiler in your report. Thank you.
java.lang.AssertionError
        at jdk.compiler/com.sun.tools.javac.util.Assert.error(Assert.java:155)
        at jdk.compiler/com.sun.tools.javac.util.Assert.check(Assert.java:46)
        at jdk.compiler/com.sun.tools.javac.comp.Flow$AliveAnalyzer.clearPendingExits(Flow.java:589)
        at jdk.compiler/com.sun.tools.javac.comp.Flow$AliveAnalyzer.visitClassDef(Flow.java:544)
        ...
```

The proposal is to clear the `returnResult` from the `Env`'s info.

This exception may affect JShell.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301580](https://bugs.openjdk.org/browse/JDK-8301580): Error recovery does not clear returnResult

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12361/head:pull/12361` \
`$ git checkout pull/12361`

Update a local copy of the PR: \
`$ git checkout pull/12361` \
`$ git pull https://git.openjdk.org/jdk pull/12361/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12361`

View PR using the GUI difftool: \
`$ git pr show -t 12361`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12361.diff">https://git.openjdk.org/jdk/pull/12361.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java [hunk 1]
```diff
@@ -5220,9 +5220,14 @@ public void visitAnnotatedType(JCAnnotatedType tree) {
     public void visitErroneous(JCErroneous tree) {
         if (tree.errs != null) {
             Env<AttrContext> errEnv = env.dup(env.tree);
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java:5222]: question: shouldn't we do: `env.dup(node, env.info.dup());` instead, that way both env's won't share the info?

Comment (lahodaj) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java]: I was a little bit on the fence on this, as some code, like `Attr.attribClass` stores and re-sets the result, while other dupes the info. Let's dupe the info then. Thanks!

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java]: I see, I didn't realize of the other pattern, not duping the info, we should probably standardize what we do at some point

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Pull request closed without merging**